index.ts

```diff
@@ -0,0 +1,552 @@
+// Exported native API for @smogon/data
+//
+// DISCLAIMER:
+//
+//   Not meant to replace Dex/BattleDex: consumers can delegate to this native API
+//   and put their own layer on top which can extend or alter the functionality.
+//
+// CONSIDERATIONS:
+//
+//    - Approach: async APIs are moved to top level so that data load can be done
+//      once and then the rest of the APIs aren't cluttered with `await` everywhere.
+//      Furthermore, individual pieces can be loaded (just `Abilities` vs. all data
+//      types) if desired.
+//    - `getByID` APIs are important for removing the `toID` call in `get` methods
+//      for when we already know through typechecking that we have an ID. String
+//      getters have been removed - callees can manually called `toID` first if they
+//      dont have an ID (which should be very rare unless its human-input).
+//    - Descriptions are UI only and can be large so aren't included in the Dex by
+//      default, but they may be loaded in bulk asynchronously for the desired
+//      languages.
+//    - Similarly, SpeciesDetails which are purely cosmetic or used by the Validator
+//      or for non-competitive reasons are offloaded. eg. learnset details are very
+//      large and not needed in many circumstances.
+//    - All other APIs are loaded together for a gen (+ mod) - it's rare a consumer
+//      cares *just* about Abilities or *just* Moves, and if they do and are in an
+//      environment where loading more is a concern they can implement their own
+//      splitting on top of the raw data files.
+//    - The aliases.json data is used internally (to implement getByID's) but is
+//      not exported.
+//    - The individual APIs (eg. Abilities) are not just Maps, as their lookup function
+//      is more sophisticated (handles aliases, caching, etc).
+//    - Internally, references can be resolved lazily through Javascript getters (though
+//      need to use `Object.defineProperty(Foo.prototype, ...)` to be able to make
+//      enumerable) to avoid object instantiation overhead and avoid creating objects for
+//      fields which are never read.
+//    - Map will need to be polyfilled or support for browsers before IE11 will need to
+//      be dropped.
+//
+//  *** TODO ***:
+//
+//   - 'Species'/'SpeciesDetails' names are problematic as they are plural
+//   - Demonstrate how PS can wrap the specific types and still extend them in OMs
+//   - All the fields in the JSON should be able to be `null` as a sentinel for removal,
+//     can implement this generically using higher order Typescript utility type
+//     functionality.
+//   - 'CommonFoo' and 'JSONFoo' naming could use bikeshedding!
+//   - Review fields to ensure optional fields are made required in the rich APIs where
+//     it makes sense to do so.
+//
+
+// core.ts
+
+type ID = '' | string & {__isID: true};
+toID(s: string): ID;
+
+type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7;
```

Comment:
I think a generation datatype would be useful. I need a two letter code (`'rb'`, `'gs'`, ...) at the minimum to support the current Smogdex URLs, even if was desired to standardize on the number, I would need to have redirects. I think it also makes sense to list the cart games covered by the generation (`['Red', 'Blue', 'Yellow']`, `['Gold', 'Silver', 'Crystal']`, ...)

This object could contain the `abilities`, `moves`, etc fields

Reply:
https://github.com/pkmn-cc/data/blob/master/src/gen.ts is what I had in a previous attempt  to handle smogon shortcodes etc.

However, I am *really* digging the idea of `Generation` being the 'God-object' compared to `Dex`. I've never liked the name `Dex`, it always felt too cutesy.

Reply:
I feel like neither `Generation` nor like `Game` are really indicative of what they actually do. `Dex` is clear, on the other hand.

Reply:
I've come to peace with `Dex`, though I dont find it clear...

Reply:
whenever someone says dex i assume they are referring to the web application. but i've always organized things by `Generation` and might be biased

Reply:
PS doesn't really even privilege `Generation`, its just a `number`. My intuition is to do things based on `Generation` (which may have a `Mod` applied), but I don't know that this is anything other than personal preference. Making `Generation` an object seems to have technical merits, but whether that object should also be the God-object is less clear technically.

I lean pretty far towards `Generation` being the entry point, but I don't know how to decide things that come down to personal preference other flipping a coin or giving in to whoever argues the loudest.

Reply:
Furthermore, mods aren't generations. It's silly to call it "Generation" when BW and BW2 and XY and XY-before-Pokébank aren't considered the same generation.

Reply:
Calling it `Mod` on the other hand makes nonzero sense, and I've seen some code that does it, but I still feel like it's not very descriptive of the purpose, it's only descriptive of what lines we've drawn for the boundaries.

`Dex` is clearest that "this is a database for pokemon, moves, items, etc".

Reply:
I feel like calling it `Generation`/`Mod` instead of `Dex` is like calling a date object `Number` instead of `Date`. `Dex` and `Date` are much more indicative of what methods you would expect to find inside it.